calliope/core.py

```diff
@@ -926,62 +907,60 @@ def read_data(self):
         data['s_init'] = xr.DataArray(s_init)
 
         # Parameters that may be defined over (x, y, t)
-        ts_sets_constraints = {'y_def_' + k: set()
-                               for k in self.config_model.timeseries_constraints}
-        self._sets = {**self._sets, **ts_sets_constraints}
-        ts_sets_costs = {'y_def_' + k[1] + '_' + k[2]: set()
-                   for k in self.config_model.timeseries_data}
-        self._sets = {**self._sets, **ts_sets_costs}
+        ts_sets = {'y_def_' + param: set()
+                   for param in self.config_model.timeseries_constraints}
+        self._sets = {**self._sets, **ts_sets}
 
         for param in self.config_model.timeseries_constraints: #constraints
             param_data = {}
-            for y in self._sets['y']:
-                # First, set up each parameter without considering
-                # potential per-location (per-x) overrides
-                k = y + '.constraints.' + param
-                option = self.get_option(k)
-
-                df = self._read_param_for_tech(param, y, time_res, option, x=None)
-
-                for x in self._sets['x']:
-                    # Check for each x whether it defines an override
-                    # that is different from the generic option, and if so,
-                    # update the dataframe
-                    option_x = self.get_option(k, x=x)
-                    if option != option_x:
-                        df.loc[:, x] = self._read_param_for_tech(param, y, time_res, option_x, x=x)
-
-                self._validate_param_df(param, y, df)  # Have all `x` been set?
-
-                param_data[y] = xr.DataArray(df, dims=['t', 'x'])
+            if 'om' in param or 'sub' in param: # cost constraints
```

Comment:
This whole `if-then` block is a bit of a mess (it already was before) and should probably be rewritten at some point, particularly now that a lot of the code is duplicated... but I guess for now it's ok, as long as it works 🙂